### ISSUE TITLE
Make the `Element` construction generic

### ISSUE THREAD

**Issue by desaikd** [amzn/ion-rust#184]
Currently the `Element` construction is separate for both `OwnedElement` and `BorrowedElement` with their own `new` methods. We need to have a generic construction based on `annotations` and `Value`. This will be useful when working with creating generic test case code and generic test cases for `Owned` and `Borrowed` . 

This implementation is trickier in terms of working with a generic `Value` implementation, which currently is an `enum` `BorrowedValue`/`OwnedValue`. It might be worth having a `new` in terms of `Self::Value` and `Self:SymbolToken` trait for this generic implementation.

**Comment by almann** [comment]
Related to #157 and #183.

**Comment by almann** [comment]
@desaikd and I talked about this at length last week, and came up with the following scheme for how to construct `Element` and `SymbolToken` generically.

Where ever possible, we should implement `From<X>` for a `...SymbolToken` and `...Element`.  We can further add this to the traits, but at a minimum it makes generic code such as `<E: Element + From<X> + ...>` possible.  To avoid boilerplate, for the common types we should add those as constraints on the `Element` and `SymbolToken` traits as base requirements for any such implementation.

```rust
trait SymbolToken
where
  Self: From<&str> + From<String> {
  ...
}
```

`clob` and `blob` are interesting cases as it is ambiguous how which one is meant for `&[u8]` or `Vec<u8>` or equivalent.  For those, we can add a builder type that is convertible from `&[u8]` and `Vec<u8>`:

```rust
trait LobBuilder
where
  Self: From<&[u8]> + From<Vec<u8>> {
  type Element: Element;
  ...
  fn into_clob(self) -> Self::Element;
  fn into_blob(self) -> Self::Element;
}
```

This makes the conversion explicit.  Likewise in `Sequence`, we should add `into_list` and `into_sexp` that are similar to the `LobBuilder` above except that `Sequence` is useful in its own right.

Some examples of the above factoring in a generic context:

```rust
fn make_stuff<E: Element>() {
  let clob_elem = E::LobBuilder::from(b"hello").into_clob();
  let blob_elem = E::LobBuilder::from(b"world).into_blob();
  ...
}
```

There is an additional trait bounds on `Element` that would require nested types to have the same element type as `Self`:

```rust
trait Element ... {
  type LobBuilder: LobBuilder<Element=Self> ...;
}
```

See [this playground link][1] for a really trivial version for this

Doing this I think will go a long way towards #136 (but may not satisfy it).

[1]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9cd1795aff74158112b495e5cd08a419

**Comment by desaikd** [comment]
This issue is closed with #199
